bftengine/src/bftengine/SigManager.hpp

```diff
@@ -68,13 +67,16 @@ class SigManager {
   void SetAggregator(std::shared_ptr<concordMetrics::Aggregator> aggregator) {
     metrics_component_.SetAggregator(aggregator);
   }
+  void setClientPublicKey(const std::string& key, PrincipalId, KeyFormat);
+
   SigManager(const SigManager&) = delete;
   SigManager& operator=(const SigManager&) = delete;
   SigManager(SigManager&&) = delete;
   SigManager& operator=(SigManager&&) = delete;
 
   concord::messages::keys_and_signatures::ClientsPublicKeys clientsPublicKeys_;
   std::string getClientsPublicKeys() {
+    std::shared_lock lock(mutex_);
```

Comment:
I don't think it's atomic.
you are replacing key by key. 
So if I want to change 10 keys, and call setClientPublicKey 5 times, then call getClientsPublicKeys, I will get 5 old keys and 5 new keys.
I think we should change the API above to setClientPublicKeys to replace all keys at once.

Reply:
Hi, after our zoom conversation I'm not sure if this comment is correct. I mean, the solution implements the interface definition to replace key by key. But on the other hand, I'm not sure what will happen if someone calls getClientsPublicKeys while replacing key by key (in the middle of the replacement). Need to think if it's a valid well-defined scenario.

Reply:
`getClientsPublicKeys()` is used in the initial client public keys publishing scenario only, so I don't see how it can create a problem.

Reply:
OK, if you take only this use case, I agree. 
But in the future, if used in other use-cases it can be a problem.
I suggest adding a comment above, stating that this might be not atomic.

Reply:
To tell the truth I don't see neither atomicity nor consistency problem here. `getClientPublicKeys()` returns keys updated to the specific moment.

Reply:
If someone calls getClientPublicKeys during an update (which is a use-case you've stated is impossible for now), it will get an output: mix of old a new keys. If this will be used later (whatever use) there will probably be a bug.

It is the responsibility of the internal implementation to prevent such inconsistencies.

A caller is not interested in a mix of old and new keys, and might be unaware of the background key-exchange process that randomly happened during the call.

Reply:
that's why there's a published_ variable

Reply:
I think we are stuck on the esoteric scenario

Reply:
I agree, this became a less practical discussion. 
In at worst case, will have to deal with a bug years/months from now and/or re-design this interface  :)